phira{,-unwrapped,-free}: init at 0.6.7

UlyssesZh opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

I wrote a patch to make the game load assets from and download data to `~/.local/share/phira` instead of the installation dir (among other patches to fix the build). Currently it seems to work without flaw on Linux.

The build on macOS cannot launch because the window creation with GL capabilities fails, but it can probably be due to I am testing in a VM (where I cannot seem to write some code myself to use GLFW to create a window, either). More testing is needed.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and others READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

UlyssesZh committed:
phira{,-unwrapped,-free}: init at 0.6.7

Bot-wxt1221 commented on pkgs/by-name/ph/phira/package.nix:
```diff
@@ -0,0 +1,99 @@
+  meta = {
+    description = "Rhythm game with custom charts and multiplayer";
+    homepage = "https://github.com/TeamFlos/phira";
+    license = lib.licenses.gpl3;
```
Is it still free with bin-src?

Bot-wxt1221 replied:
Let's use only first.

UlyssesZh replied:
Ah it is GPL-3.0-only as can be seen in Cargo.toml. I didn't notice it at first.

Bot-wxt1221 commented on pkgs/by-name/ph/phira/package.nix:
```diff
@@ -0,0 +1,99 @@
+    ]);
+
+  # for crate coreaudio-sys
+  LIBCLANG_PATH = lib.optionalString stdenv.hostPlatform.isDarwin "${lib.getLib llvmPackages.libclang}/lib";
```
use rustPlatform.bindgenHook

Bot-wxt1221 commented on pkgs/by-name/ph/phira/package.nix:
```diff
@@ -0,0 +1,99 @@
+    description = "Rhythm game with custom charts and multiplayer";
+    homepage = "https://github.com/TeamFlos/phira";
+    license = lib.licenses.gpl3;
+    maintainers = with lib.maintainers; [ ulysseszhan ];
```
```suggestion
    maintainers = with lib.maintainers; [ ulysseszhan ];
    changelog = "https://github.com/TeamFlos/phira/releases/tag/v${version}";
```

Bot-wxt1221 commented on pkgs/by-name/ph/phira/package.nix:
```diff
@@ -0,0 +1,99 @@
+    ./assets.patch
+  ];
+
+  cargoHash = "sha256-6mRb3M56G20fA+px1cZyrGpel0v54qoVAQK2ZgTzkmI=";
```
useFetchCargoVendor then update hash.

Bot-wxt1221 reviewed (commented):
patch lgtm

Bot-wxt1221 commented on pkgs/by-name/ph/phira/package.nix:
```diff
@@ -0,0 +1,99 @@
+    hash = "sha256-iVwfuqU/jTPULjCtaPOcnYLNJZzEzFjvM0PqEGULLCw=";
+  };
+in
+rustPlatform.buildRustPackage {
```
I suggest finalAttrs instead of let in

UlyssesZh replied:
#315337, but yes I think `finallAttrs` is better. I will change it.

Bot-wxt1221 reviewed (approved):
All thing worked!

Bot-wxt1221 reviewed (commented):
Package work for me with x11

<img width="490" height="306" alt="Image" src="https://github.com/user-attachments/assets/f47f457c-3e44-4676-ad82-e3d20052b571" />

Bot-wxt1221 commented:
cachix itself is broken.

Bot-wxt1221 commented:
Let's unfree. 

https://github.com/TeamFlos/phira/issues/578#issuecomment-3092142612

UlyssesZh commented:
He didn't understand the question at first. He thought you were talking about some source codes.

Bot-wxt1221 commented:
Could there be a bin package?

https://github.com/TeamFlos/phira/issues/298

Aleksanaa commented on pkgs/by-name/ph/phira/assets.patch:
```diff
@@ -0,0 +1,19 @@
+ }
+ 
+ pub fn init_assets() {
++    if let Ok(root_path) = std::env::var("PHIRA_ROOT") {
```
It would be better to follow XDG specification here: `$XDG_DATA_HOME` -> `$HOME/.local/share` -> `~/.local/share` (https://specifications.freedesktop.org/basedir-spec/latest/)

UlyssesZh replied:
I don't want to change the default behavior when no environment variables are set.

UlyssesZh replied:
The wrapper script now is aware of `$XDG_DATA_HOME`.

Aleksanaa commented on pkgs/by-name/ph/phira/package.nix:
```diff
@@ -0,0 +1,104 @@
+      makeWrapper $out/bin/phira-main $out/bin/phira \
+        ${lib.optionalString stdenv.hostPlatform.isLinux "--suffix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ libGL ]}"} \
+        --run '
+          export PHIRA_ROOT=''${PHIRA_ROOT-"$HOME/.local/share/phira"}
```
Don't think putting the logic here is a good one. And without it we can make use of `makeBinaryWrapper` which should be faster.

UlyssesZh replied:
I don't want to maintain a patch with too much logic. The performance impact from `makeWrapper` is very minimal.

UlyssesZh commented on pkgs/by-name/ph/phira/package.nix:
```diff
@@ -0,0 +1,104 @@
+    let
+      # There are some assets that the developer put in gitignore but are required to launch the game,
+      # so we grab them from the binary release.
+      bin-src = fetchurl {
```
It is now overridable as a package argument.

UlyssesZh commented:
I separated it into two packages `phira` and `phira-unwrapped`. The unwrapped package does not have assets but is free. This will fix the binary cache problem.

UlyssesZh commented:
A desktop file is now added.

`phira-main` and `phira-monitor` are now wrapped so that one can run them directly.

I added the package `phira-free`, which has a pure-color image as background and a free font (Ubuntu regular). The font does not support CJK, so it is a bit bad.

SuperSandro2000 commented on pkgs/by-name/ph/phira-unwrapped/ffmpeg.patch:
Would be nice to upstream this
